### ISSUE TITLE
training problem

### ISSUE THREAD

**Issue by aopolin-lv** [lxa9867/ImageFolder#24]
Hi, this repo is excellent.

I am trying to reimplement the experiment with the config `MSVR10P2-4096.yaml` but set the `enc_type: cnn` and `dec_type: cnn`. However, there are bugs:

- In the line 146 of `xqgan_model.py`, the code can not run. The decoder has no attribute of embed_dim.
```self.post_quant_conv = nn.Conv2d(config.codebook_embed_dim * self.product_quant, self.decoder.embed_dim, 1)```

- Also, when I fix bug above by deleting the bugged line, in the line 94 of `quant.py`, it raises errors, maybe the mismatch of dimension.
```idx_N = torch.argmax(rest_NC @ F.normalize(self.embedding.weight.data.T, dim=0), dim=1)```

**Comment by qiuk2** [comment]
Thanks for your interest!

Currently, we do not support pq=2 for CNN-based tokenizers in XQGAN, as we found it does not effectively disentangle semantic and detail information.

If you really need to use a CNN-based tokenizer, we recommend setting pq=1, which is more stable in this case.

**Comment by aopolin-lv** [comment]
> Thanks for your interest!
> 
> Currently, we do not support pq=2 for CNN-based tokenizers in XQGAN, as we found it does not effectively disentangle semantic and detail information.
> 
> If you really need to use a CNN-based tokenizer, we recommend setting pq=1, which is more stable in this case.

Thank you for the response. I have another question: if I want to train a vqvae model with my own data (~20M images), should I load the ckpt trained on the imagenet or from scratch? If loading the ckpt, what range of lr is suitable? Thanks again!

**Comment by lxa9867** [comment]
If you are training on ~20M images, I would recommend training from scratch (but loading dino/clip weights for transformer initialization). If the dataset is text-rich, it's better to set the codebook size to a large value.

We did not ablate the lr for training from scratch with large-scale dataset but the default setting for imagenet should make the model to converge eventually in our experiment.

**Comment by aopolin-lv** [comment]
Thanks a lot for the advice! Just working on training my model at the moment.

**Comment by aopolin-lv** [comment]
Hello, I would like to ask about the utilization rate of the codebook. I am using ImageNet as the training set to train vq-vae, and after more than 180 epochs of training, its FID can drop to about 0.8, but the usage is only 40%+, and the effect is very poor. How can I do to increase the usage?

My experimental setting is that pq=1 is set, and trained according to the var setting. batch_size=1024, 48 a100 training, and other parameters are default.

<img width="1561" alt="Image" src="https://github.com/user-attachments/assets/f8b62fe3-a158-49e1-832a-08b182f2438c" />
<img width="508" alt="Image" src="https://github.com/user-attachments/assets/ecfaf563-f315-4d8b-b396-7af56e795c75" />
<img width="510" alt="Image" src="https://github.com/user-attachments/assets/656f6258-8017-4d30-b5a9-e50b4a2bd579" />

<img width="1558" alt="Image" src="https://github.com/user-attachments/assets/e5351348-0207-416c-b61e-cc809c0e515a" />

**Comment by qiuk2** [comment]
May I ask your training config? Also, can you report your codebook usage for each layer?

**Comment by aopolin-lv** [comment]
The config is as follows:
```
data_path: datasets/ImageNet-1k/raw/ImageNet-1K/train
val_data_path: datasets/ImageNet-1k/raw/ImageNet-1K/val
save_best: true
image_size: 256
vq_model: VQ-16
enc_type: cnn
dec_type: cnn
semantic_guide: none
# detail_guide: none
global_batch_size: 1024
epochs: 200
lr_scheduler: cosine
lr: 3e-5
#max_grad_norm: 1.0
#reconstruction_weight: 4.0
codebook_embed_dim: 32
codebook_size: 4096
abs_pos_embed: true
product_quant: 1
ema: true
codebook_drop: 0.1
encoder_model: vit_base_patch14_dinov2.lvd142m
decoder_model: vit_base_patch14_dinov2.lvd142m
num_latent_tokens: 680
v_patch_nums: [1, 2, 3, 4, 5, 6, 8, 10, 13, 16]
half_sem: true
start_drop: 3
ckpt_every: 10000
sem_loss_weight: 0.1
enc_tuning_method: full
lecam_loss_weight: 0.001
weight_decay: 0.0
disc_weight_decay: 0.0005
disc_epoch_start: 56
disc_type: dinodisc
aug_fade_steps: 0
disc_adaptive_weight: true
```

The training log is:

<img width="1568" alt="Image" src="https://github.com/user-attachments/assets/4c7f4648-76de-4a3d-9402-fd85d6b64cfb" />

**Comment by aopolin-lv** [comment]
At the same time, when I tried to fine-tune the pre-trained ckpt on my data, the utilization rate quickly dropped from 86% to 56% in the first 100 steps. I think a utilization rate of 56% is unreasonable for training autoregressive models. Training is still ongoing. I don't know if it can increase to above 80%.

The main setting of finetune period is batch_sizez=1024, lr=1e-5, epoch=40, len(dataset)=~10M, 128gpus, and others are default.

<img width="393" alt="Image" src="https://github.com/user-attachments/assets/a52fda17-349f-4d02-9cb4-43372309c6f4" />

**Comment by lxa9867** [comment]
Hi, 

I would like to clarify something:

1.  The codebook utilization is 100%. In the residual quantization, each rq layer is supposed to use different part of the codebook. So it shouldn't be just summed up for the final codebook utilization. Our implementation is just for convenience. Sorry for the confusion.
2. Empirically, a larger utilization rate in the first several layers may help better generation, i.e., have more diversity in the first several AR steps. I would suggest to incorporate a entropy loss if your current experiment doesn't yield a satisfactory result. I would like to note that the entropy loss do help with the utilization rate but it is difficult to adjust as the higher utilization does not positively correlated to a better gFID.
3. A semantic distillation may help achieve a better utilization rate. If you are using a generator smaller than 3B, I would suggest add a distillation loss. If not, you may not add it I think. In addition, if you add the noise perturbation, you may also have a better FID (verified by three community feedback).

Pls lmk if you have other questions. Thanks.
